src/app/games/hive/HivePieceBehaviour.ts

```diff
@@ -0,0 +1,298 @@
+import { Coord } from 'src/app/jscaip/Coord';
+import { HexaDirection } from 'src/app/jscaip/HexaDirection';
+import { HexagonalUtils } from 'src/app/jscaip/HexagonalUtils';
+import { MGPFallible } from 'src/app/utils/MGPFallible';
+import { MGPOptional } from 'src/app/utils/MGPOptional';
+import { MGPSet } from 'src/app/utils/MGPSet';
+import { Utils } from 'src/app/utils/utils';
+import { HiveFailure } from './HiveFailure';
+import { HiveMoveCoordToCoord, HiveMoveSpider } from './HiveMove';
+import { HivePiece, HivePieceKind, HivePieceStack } from './HivePiece';
+import { HiveState } from './HiveState';
+
+export abstract class HivePieceBehaviour {
+
+    private static INSTANCES: MGPOptional<Record<HivePieceKind, HivePieceBehaviour>> = MGPOptional.empty();
+
+    public static from(piece: HivePiece): HivePieceBehaviour {
+        if (HivePieceBehaviour.INSTANCES.isAbsent()) {
+            HivePieceBehaviour.INSTANCES = MGPOptional.of({
+                'QueenBee': HivePieceBehaviourQueenBee.get(),
+                'Beetle': HivePieceBehaviourBeetle.get(),
+                'Grasshopper': HivePieceBehaviourGrasshopper.get(),
+                'Spider': HivePieceBehaviourSpider.get(),
+                'SoldierAnt': HivePieceBehaviourSoldierAnt.get(),
+            });
+        }
+        return HivePieceBehaviour.INSTANCES.get()[piece.kind];
+    }
+
+    protected destinationIsEmpty(move: HiveMoveCoordToCoord, state: HiveState): boolean {
+        return state.getAt(move.end).isEmpty();
+    }
+
+    protected canSlide(state: HiveState, start: Coord, end: Coord): boolean {
+        // For the piece to slide from start to end,
+        // one of the two common neighbors of start and coord should be empty
+        const startNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(start));
+        const endNeighbors: MGPSet<Coord> = new MGPSet(HexagonalUtils.neighbors(end));
+        const commonNeighbors: MGPSet<Coord> = startNeighbors.intersect(endNeighbors);
+        for (const neighbor of commonNeighbors) {
+            if (state.getAt(neighbor).isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public abstract moveValidity(move: HiveMoveCoordToCoord, state: HiveState): MGPFallible<void>
+
+    public abstract getPossibleMoves(coord: Coord, state: HiveState): HiveMoveCoordToCoord[]
+
+}
+
+export class HivePieceBehaviourQueenBee extends HivePieceBehaviour {
```

Comment:
QueenBeeBehaviour ?
Btw, Behavior in american, amaright ?

Reply:
toutafé pour Behavior, par contre pas fan de pas préfixer par Hive